frontend/ui/plugin/switch_plugin.lua

```diff
@@ -0,0 +1,109 @@
+--[[--
+SwitchPlugin creates a plugin with a switch to enable or disable it.
+See spec/unit/switch_plugin_spec.lua for the usage.
+]]
+
+local ConfirmBox = require("ui/widget/confirmbox")
+local DataStorage = require("datastorage")
+local LuaSettings = require("luasettings")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local logger = require("logger")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+local SwitchPlugin = WidgetContainer:new()
+
+function SwitchPlugin:extend(o)
+    o = o or {}
+    setmetatable(o, self)
+    self.__index = self
+    return o
+end
+
+function SwitchPlugin:new(o)
+    o = self:extend(o)
+    assert(type(o.name) == "string", "name is required");
+    o.settings = LuaSettings:open(DataStorage:getSettingsDir() .. "/" .. o.name .. ".lua")
+    o.settings_id = 0
+    SwitchPlugin._init(o)
+    return o
+end
+
+function SwitchPlugin:_init()
+    if self.default_enable then
+        self.enabled = self.settings:nilOrTrue("enable")
+    else
+        self.enabled = not self.settings:nilOrFalse("enable")
+    end
+    self.settings_id = self.settings_id + 1
+    logger.dbg("SwitchPlugin:_init() self.enabled: ", self.enabled, " with id ", self.settings_id)
+    if self.enabled then
+        self:_start()
+    else
+        self:_stop()
+    end
+end
+
+function SwitchPlugin:flipSetting()
+    if self.default_enable then
+        self.settings:flipNilOrTrue("enable")
+    else
+        self.settings:flipNilOrFalse("enable")
+    end
+    self:_init()
+end
+
+function SwitchPlugin:onFlushSettings()
+    self.settings:flush()
+end
+
+--- Show a ConfirmBox to ask for enabling or disabling this plugin.
+function SwitchPlugin:_showConfirmBox()
+    UIManager:show(ConfirmBox:new{
+        text = self._confirmMessage() ..
+               T(_("Do you want to %1 it?"),
```

Comment:
See my comment at https://github.com/koreader/koreader/pull/3078#discussion_r134671887

> Is that safe for localization? It might be better to just write two full sentences, even if that seems redundant to a programmer.

Reply:
Writing into two sentences is not the right thing for localization. Some languages do not have the same syntax structure as English, so %1 may be placed before or after other elements. E.g. in Chinese, we usually say,
是否要%1？
%1 = 禁用 (disable) or 启用 (enable)
Because Chinese is not a syntax-first language, "it" is usually ignored. The context in the _confirmMessage() provides enough information to indicate what would be disabled.
(P.S. Syntax in Chinese is more complex than the example above, sometimes we can randomly put predicate before or after the object without breaking the meaning.)

Reply:
> Writing into two sentences is not the right thing for localization. Some languages do not have the same syntax structure as English, so %1 may be placed before or after other elements.

But that's my whole point! In translating the *whole* sentence you're not dependent on random words like *enable* and *disable* being exactly the same as those that would actually fit into the sentence.

Reply:
Sorry, I misunderstood your meaning :(

Reply:
No problem at all. :)